File: golang/676_magic_dictionary/magic_dictionary.go
/*
 * @Date: 2022-07-11
 * @LastEditors: mengsenwang mengsen_wang@163.com
 * @LastEditTime: 2022-07-11
 * @FilePath: /algorithm/676_magic_dictionary/magic_dictionary.go
 */

package main

type trie struct {
	children   [26]*trie
	isFinished bool
}

type MagicDictionary struct {
	*trie
}

func Constructor() MagicDictionary {
	return MagicDictionary{&trie{}}
}

func (d *MagicDictionary) BuildDict(dictionary []string) {
	for _, word := range dictionary {
		cur := d.trie
		for _, c := range word {
			c -= 'a'
			if cur.children[c] == nil {
				cur.children[c] = &trie{}
			}
			cur = cur.children[c]
		}
		cur.isFinished = true
	}
}

func dfs(node *trie, searchWord string, modified bool) bool {
	if searchWord == "" {
		return modified && node.isFinished
	}
	c := searchWord[0] - 'a'
	if node.children[c] != nil && dfs(node.children[c], searchWord[1:], modified) {
		return true
	}
	if !modified {
		for i, child := range node.children {
			if i != int(c) && child != nil && dfs(child, searchWord[1:], true) {
				return true
			}
		}
	}
	return false
}

func (d *MagicDictionary) Search(searchWord string) bool {
	return dfs(d.trie, searchWord, false)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	m := Constructor()
	m.BuildDict([]string{"hello", "leetcode"})
	assert(m.Search("hello") == false)
	assert(m.Search("hhllo") == true)
	assert(m.Search("hell") == false)
	assert(m.Search("leetcoded") == false)
}

File: cpp/676_magic_dictionary/magic_dictionary.cpp
/*
 * @Date: 2022-07-11
 * @LastEditors: mengsenwang mengsen_wang@163.com
 * @LastEditTime: 2022-07-11
 * @FilePath: /algorithm/676_magic_dictionary/magic_dictionary.cpp
 */

#include <algorithm>
#include <cassert>
#include <functional>
#include <string>
#include <vector>

using namespace std;

struct Trie {
  bool is_finished;
  Trie* child[26];

  Trie() {
    is_finished = false;
    fill(begin(child), end(child), nullptr);
  }
};

class MagicDictionary {
 public:
  MagicDictionary() { root = new Trie(); }

  void buildDict(vector<string> dictionary) {
    for (auto&& word : dictionary) {
      Trie* cur = root;
      for (char ch : word) {
        int idx = ch - 'a';
        if (!cur->child[idx]) {
          cur->child[idx] = new Trie();
        }
        cur = cur->child[idx];
      }
      cur->is_finished = true;
    }
  }

  bool search(string searchWord) {
    function<bool(Trie*, int, bool)> dfs = [&](Trie* node, int pos, bool modified) {
      if (pos == searchWord.size()) {
        return modified && node->is_finished;
      }
      int idx = searchWord[pos] - 'a';
      if (node->child[idx]) {
        if (dfs(node->child[idx], pos + 1, modified)) {
          return true;
        }
      }
      if (!modified) {
        for (int i = 0; i < 26; ++i) {
          if (i != idx && node->child[i]) {
            if (dfs(node->child[i], pos + 1, true)) {
              return true;
            }
          }
        }
      }
      return false;
    };

    return dfs(root, 0, false);
  }

 private:
  Trie* root;
};

int main() {
  MagicDictionary m{};
  m.buildDict(vector<string>{{"hello", "leetcode"}});
  assert(m.search("hello") == false);
  assert(m.search("hhllo") == true);
  assert(m.search("hell") == false);
  assert(m.search("leetcoded") == false);
}